lib/travis/build/stages.rb

```diff
@@ -58,24 +67,54 @@ def initialize(script, sh, config)
       end
 
       def run
-        stages.each_slice(2) do |type, names|
-          names.each { |name| run_stage(type, name) }
+        run_stage(:builtin, :header)
+
+        STAGES.each do |stage|
+          run_stage(stage.type, stage.name)
         end
-      end
 
-      def stages
-        script.debug_build_via_api? ? STAGES_DEBUG : STAGES
+        sh.raw "source $HOME/.build_stages"
+
+        STAGES.each do |stage|
+          case stage.run_in_debug
+          when :always
+            sh.raw "run_stage_#{stage.name}"
+          when true
+            sh.raw "run_stage_#{stage.name}" if debug_build?
+          when false
+            sh.raw "run_stage_#{stage.name}" unless debug_build?
+          end
+        end
       end
 
       def run_stage(type, name)
-        type = :builtin if fallback?(type, name)
-        stage = self.class.const_get(type.to_s.camelize).new(script, name)
-        stage.run
+        wrap_in_func(name) do
+          type = :builtin if fallback?(type, name)
+          stage = self.class.const_get(type.to_s.camelize).new(script, name)
+          commands = stage.run
+        end
       end
 
       def fallback?(type, name)
         type == :custom && !config[name]
       end
+
+      def wrap_in_func(stage)
+        if stage == :header
+          yield
+        else
+          sh.raw "cat <<'EOFUNC' >>$HOME/.build_stages"
```

Comment:
With the quotation marks in the here-doc marker, no interpretation is performed. Example 19.7 in http://www.tldp.org/LDP/abs/html/here-docs.html.